File: src/bindings/tests/benchmark_tests.py
#!/usr/bin/env python3
"""
Comprehensive benchmark tests for Python bindings of OrderBook-rs
Performance testing and regression detection
"""
import time
import statistics
import sys
import os
import json
import gc
from datetime import datetime
from typing import List, Dict, Any

# Add the built module to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..', 'target', 'release'))

try:
    from orderbook_rs import PyOrderBook
except ImportError as e:
    print(f"Failed to import PyOrderBook: {e}")
    print("Make sure to build with: maturin develop --features python")
    sys.exit(1)


class BenchmarkResult:
    """Container for benchmark results."""
    
    def __init__(self, name: str, operations: int, duration: float, memory_usage: float = 0.0):
        self.name = name
        self.operations = operations
        self.duration = duration
        self.ops_per_second = operations / duration if duration > 0 else 0
        self.memory_usage = memory_usage
        self.timestamp = datetime.now()
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'name': self.name,
            'operations': self.operations,
            'duration': self.duration,
            'ops_per_second': self.ops_per_second,
            'memory_usage': self.memory_usage,
            'timestamp': self.timestamp.isoformat()
        }


class OrderBookBenchmark:
    """Comprehensive benchmark suite for OrderBook operations."""
    
    def __init__(self):
        self.results: List[BenchmarkResult] = []
        self.warmup_iterations = 1000
        self.benchmark_iterations = 10000
    
    def warmup(self, orderbook: PyOrderBook):
        """Warm up the JIT and caches."""
        print("🔥 Warming up...")
        for i in range(self.warmup_iterations):
            price = 50000 + (i % 100)
            side = "Buy" if i % 2 == 0 else "Sell"
            orderbook.add_limit_order(price, 100, side, "GTC")
    
    def benchmark_single_order_insertion(self) -> BenchmarkResult:
        """Benchmark single order insertion performance."""
        print("📊 Benchmarking single order insertion...")
        
        orderbook = PyOrderBook("BENCHMARK_SINGLE")
        self.warmup(orderbook)
        
        gc.collect()
        start_time = time.perf_counter()
        
        for i in range(self.benchmark_iterations):
            price = 50000 + (i % 1000)
            side = "Buy" if i % 2 == 0 else "Sell"
            orderbook.add_limit_order(price, 100, side, "GTC")
        
        end_time = time.perf_counter()
        duration = end_time - start_time
        
        result = BenchmarkResult("Single Order Insertion", self.benchmark_iterations, duration)
        self.results.append(result)
        return result
    
    def benchmark_batch_order_insertion(self) -> BenchmarkResult:
        """Benchmark batch order insertion performance."""
        print("📊 Benchmarking batch order insertion...")
        
        orderbook = PyOrderBook("BENCHMARK_BATCH")
        batch_size = 1000
        batch_count = 10
        
        gc.collect()
        start_time = time.perf_counter()
        
        for batch in range(batch_count):
            for i in range(batch_size):
                price = 50000 + (i % 100)
                side = "Buy" if i % 2 == 0 else "Sell"
                orderbook.add_limit_order(price, 100, side, "GTC")
        
        end_time = time.perf_counter()
        duration = end_time - start_time
        total_operations = batch_size * batch_count
        
        result = BenchmarkResult("Batch Order Insertion", total_operations, duration)
        self.results.append(result)
        return result
    
    def benchmark_price_queries(self) -> BenchmarkResult:
        """Benchmark price query performance."""
        print("📊 Benchmarking price queries...")
        
        orderbook = PyOrderBook("BENCHMARK_QUERIES")
        
        # Add some orders first
        for i in range(1000):
            price = 50000 + (i % 100)
            side = "Buy" if i % 2 == 0 else "Sell"
            orderbook.add_limit_order(price, 100, side, "GTC")
        
        query_count = 100000
        
        gc.collect()
        start_time = time.perf_counter()
        
        for _ in range(query_count):
            bid = orderbook.best_bid()
            ask = orderbook.best_ask()
        
        end_time = time.perf_counter()
        duration = end_time - start_time
        
        result = BenchmarkResult("Price Queries", query_count * 2, duration)
        self.results.append(result)
        return result
    
    def benchmark_mixed_operations(self) -> BenchmarkResult:
        """Benchmark mixed operations (realistic trading scenario)."""
        print("📊 Benchmarking mixed operations...")
        
        orderbook = PyOrderBook("BENCHMARK_MIXED")
        operation_count = 5000
        
        gc.collect()
        start_time = time.perf_counter()
        
        for i in range(operation_count):
            # 70% order insertion, 30% price queries
            if i % 10 < 7:
                price = 50000 + (i % 100)
                side = "Buy" if i % 2 == 0 else "Sell"
                orderbook.add_limit_order(price, 100, side, "GTC")
            else:
                bid = orderbook.best_bid()
                ask = orderbook.best_ask()
        
        end_time = time.perf_counter()
        duration = end_time - start_time
        
        result = BenchmarkResult("Mixed Operations", operation_count, duration)
        self.results.append(result)
        return result
    
    def benchmark_concurrent_simulation(self) -> BenchmarkResult:
        """Simulate concurrent access patterns."""
        print("📊 Benchmarking concurrent simulation...")
        
        import threading
        import concurrent.futures
        
        orderbook = PyOrderBook("BENCHMARK_CONCURRENT")
        operations_per_thread = 1000
        thread_count = 4
        
        def worker_thread(thread_id: int):
            for i in range(operations_per_thread):
                price = 50000 + thread_id * 1000 + (i % 100)
                side = "Buy" if i % 2 == 0 else "Sell"
                orderbook.add_limit_order(price, 100, side, "GTC")
        
        gc.collect()
        start_time = time.perf_counter()
        
        with concurrent.futures.ThreadPoolExecutor(max_workers=thread_count) as executor:
            futures = [executor.submit(worker_thread, i) for i in range(thread_count)]
            concurrent.futures.wait(futures)
        
        end_time = time.perf_counter()
        duration = end_time - start_time
        total_operations = operations_per_thread * thread_count
        
        result = BenchmarkResult("Concurrent Simulation", total_operations, duration)
        self.results.append(result)
        return result
    
    def benchmark_memory_usage(self) -> BenchmarkResult:
        """Benchmark memory usage patterns."""
        print("📊 Benchmarking memory usage...")
        
        import psutil
        import os
        
        process = psutil.Process(os.getpid())
        initial_memory = process.memory_info().rss / 1024 / 1024  # MB
        
        orderbook = PyOrderBook("BENCHMARK_MEMORY")
        order_count = 50000
        
        gc.collect()
        start_time = time.perf_counter()
        
        for i in range(order_count):
            price = 50000 + (i % 1000)
            side = "Buy" if i % 2 == 0 else "Sell"
            orderbook.add_limit_order(price, 100, side, "GTC")
        
        end_time = time.perf_counter()
        duration = end_time - start_time
        
        final_memory = process.memory_info().rss / 1024 / 1024  # MB
        memory_usage = final_memory - initial_memory
        
        result = BenchmarkResult("Memory Usage", order_count, duration, memory_usage)
        self.results.append(result)
        return result
    
    def benchmark_latency_distribution(self) -> BenchmarkResult:
        """Benchmark latency distribution for individual operations."""
        print("📊 Benchmarking latency distribution...")
        
        orderbook = PyOrderBook("BENCHMARK_LATENCY")
        operation_count = 10000
        latencies = []
        
        # Warm up
        for i in range(100):
            orderbook.add_limit_order(50000 + i, 100, "Buy", "GTC")
        
        gc.collect()
        
        for i in range(operation_count):
            start = time.perf_counter()
            price = 50000 + (i % 1000)
            side = "Buy" if i % 2 == 0 else "Sell"
            orderbook.add_limit_order(price, 100, side, "GTC")
            end = time.perf_counter()
            latencies.append((end - start) * 1000000)  # microseconds
        
        # Calculate statistics
        mean_latency = statistics.mean(latencies)
        median_latency = statistics.median(latencies)
        p95_latency = sorted(latencies)[int(0.95 * len(latencies))]
        p99_latency = sorted(latencies)[int(0.99 * len(latencies))]
        
        print(f"    Mean latency: {mean_latency:.2f} μs")
        print(f"    Median latency: {median_latency:.2f} μs")
        print(f"    95th percentile: {p95_latency:.2f} μs")
        print(f"    99th percentile: {p99_latency:.2f} μs")
        
        total_duration = sum(latencies) / 1000000  # convert back to seconds
        result = BenchmarkResult("Latency Distribution", operation_count, total_duration)
        self.results.append(result)
        return result
    
    def run_all_benchmarks(self):
        """Run all benchmark tests."""
        print("🚀 Starting OrderBook Python Binding Benchmarks")
        print("=" * 60)
        
        benchmarks = [
            self.benchmark_single_order_insertion,
            self.benchmark_batch_order_insertion,
            self.benchmark_price_queries,
            self.benchmark_mixed_operations,
            self.benchmark_concurrent_simulation,
            self.benchmark_memory_usage,
            self.benchmark_latency_distribution
        ]
        
        for benchmark in benchmarks:
            try:
                result = benchmark()
                print(f"✅ {result.name}: {result.ops_per_second:,.0f} ops/sec")
                if result.memory_usage > 0:
                    print(f"    Memory usage: {result.memory_usage:.2f} MB")
            except Exception as e:
                print(f"❌ {benchmark.__name__} failed: {e}")
        
        self.print_summary()
        self.save_results()
    
    def print_summary(self):
        """Print benchmark summary."""
        print("\n" + "=" * 60)
        print("📊 BENCHMARK SUMMARY")
        print("=" * 60)
        
        for result in self.results:
            print(f"{result.name:.<30} {result.ops_per_second:>15,.0f} ops/sec")
        
        if self.results:
            total_ops = sum(r.operations for r in self.results)
            total_time = sum(r.duration for r in self.results)
            average_throughput = total_ops / total_time if total_time > 0 else 0
            
            print("-" * 60)
            print(f"{'Overall Average':.<30} {average_throughput:>15,.0f} ops/sec")
            print(f"{'Total Operations':.<30} {total_ops:>15,} ops")
            print(f"{'Total Time':.<30} {total_time:>15.3f} sec")
    
    def save_results(self):
        """Save benchmark results to JSON file."""
        results_file = "benchmark_results_python.json"
        
        results_data = {
            'timestamp': datetime.now().isoformat(),
            'platform': 'Python',
            'results': [r.to_dict() for r in self.results]
        }
        
        with open(results_file, 'w') as f:
            json.dump(results_data, f, indent=2)
        
        print(f"\n📄 Results saved to {results_file}")
    
    def compare_with_baseline(self, baseline_file: str = "benchmark_baseline.json"):
        """Compare results with baseline performance."""
        if not os.path.exists(baseline_file):
            print(f"⚠️  Baseline file {baseline_file} not found")
            return
        
        try:
            with open(baseline_file, 'r') as f:
                baseline_data = json.load(f)
            
            print(f"\n📈 PERFORMANCE COMPARISON vs {baseline_file}")
            print("=" * 60)
            
            baseline_results = {r['name']: r for r in baseline_data['results']}
            
            for result in self.results:
                if result.name in baseline_results:
                    baseline_ops = baseline_results[result.name]['ops_per_second']
                    current_ops = result.ops_per_second
                    
                    if baseline_ops > 0:
                        improvement = ((current_ops - baseline_ops) / baseline_ops) * 100
                        status = "📈" if improvement > 0 else "📉" if improvement < -5 else "📊"
                        print(f"{result.name:.<30} {status} {improvement:>8.1f}%")
                    else:
                        print(f"{result.name:.<30} ❓ No baseline data")
                else:
                    print(f"{result.name:.<30} 🆕 New benchmark")
                    
        except Exception as e:
            print(f"❌ Error comparing with baseline: {e}")


def main():
    """Main benchmark execution."""
    try:
        benchmark = OrderBookBenchmark()
        benchmark.run_all_benchmarks()
        benchmark.compare_with_baseline()
        
        print("\n🎉 Python binding benchmarks completed successfully!")
        return 0
        
    except KeyboardInterrupt:
        print("\n⏹️  Benchmark interrupted by user")
        return 1
    except Exception as e:
        print(f"❌ Benchmark failed: {e}")
        return 1


if __name__ == '__main__':
    sys.exit(main())

File: src/bindings/tests/benchmark_tests.js
#!/usr/bin/env node
/**
 * Comprehensive benchmark tests for Node.js bindings of OrderBook-rs
 * Performance testing and regression detection
 */

const fs = require('fs');
const path = require('path');

// Import the built module
let JsOrderBook;
try {
    JsOrderBook = require('../../../index.js').JsOrderBook;
} catch (error) {
    console.error('Failed to import JsOrderBook:', error.message);
    console.error('Make sure to build with: npm run build');
    process.exit(1);
}

/**
 * Container for benchmark results
 */
class BenchmarkResult {
    constructor(name, operations, duration, memoryUsage = 0) {
        this.name = name;
        this.operations = operations;
        this.duration = duration;
        this.opsPerSecond = operations / duration;
        this.memoryUsage = memoryUsage;
        this.timestamp = new Date().toISOString();
    }

    toObject() {
        return {
            name: this.name,
            operations: this.operations,
            duration: this.duration,
            opsPerSecond: this.opsPerSecond,
            memoryUsage: this.memoryUsage,
            timestamp: this.timestamp
        };
    }
}

/**
 * Comprehensive benchmark suite for OrderBook operations
 */
class OrderBookBenchmark {
    constructor() {
        this.results = [];
        this.warmupIterations = 1000;
        this.benchmarkIterations = 10000;
    }

    /**
     * Warm up the V8 engine and caches
     */
    warmup(orderbook) {
        console.log('🔥 Warming up...');
        for (let i = 0; i < this.warmupIterations; i++) {
            const price = 50000 + (i % 100);
            const side = i % 2 === 0 ? 'Buy' : 'Sell';
            orderbook.addLimitOrder(price, 100, side, 'GTC');
        }
    }

    /**
     * Benchmark single order insertion performance
     */
    benchmarkSingleOrderInsertion() {
        console.log('📊 Benchmarking single order insertion...');
        
        const orderbook = new JsOrderBook('BENCHMARK_SINGLE');
        this.warmup(orderbook);
        
        if (global.gc) global.gc();
        
        const startTime = process.hrtime.bigint();
        
        for (let i = 0; i < this.benchmarkIterations; i++) {
            const price = 50000 + (i % 1000);
            const side = i % 2 === 0 ? 'Buy' : 'Sell';
            orderbook.addLimitOrder(price, 100, side, 'GTC');
        }
        
        const endTime = process.hrtime.bigint();
        const duration = Number(endTime - startTime) / 1000000000; // Convert to seconds
        
        const result = new BenchmarkResult('Single Order Insertion', this.benchmarkIterations, duration);
        this.results.push(result);
        return result;
    }

    /**
     * Benchmark batch order insertion performance
     */
    benchmarkBatchOrderInsertion() {
        console.log('📊 Benchmarking batch order insertion...');
        
        const orderbook = new JsOrderBook('BENCHMARK_BATCH');
        const batchSize = 1000;
        const batchCount = 10;
        
        if (global.gc) global.gc();
        
        const startTime = process.hrtime.bigint();
        
        for (let batch = 0; batch < batchCount; batch++) {
            for (let i = 0; i < batchSize; i++) {
                const price = 50000 + (i % 100);
                const side = i % 2 === 0 ? 'Buy' : 'Sell';
                orderbook.addLimitOrder(price, 100, side, 'GTC');
            }
        }
        
        const endTime = process.hrtime.bigint();
        const duration = Number(endTime - startTime) / 1000000000;
        const totalOperations = batchSize * batchCount;
        
        const result = new BenchmarkResult('Batch Order Insertion', totalOperations, duration);
        this.results.push(result);
        return result;
    }

    /**
     * Benchmark price query performance
     */
    benchmarkPriceQueries() {
        console.log('📊 Benchmarking price queries...');
        
        const orderbook = new JsOrderBook('BENCHMARK_QUERIES');
        
        // Add some orders first
        for (let i = 0; i < 1000; i++) {
            const price = 50000 + (i % 100);
            const side = i % 2 === 0 ? 'Buy' : 'Sell';
            orderbook.addLimitOrder(price, 100, side, 'GTC');
        }
        
        const queryCount = 100000;
        
        if (global.gc) global.gc();
        
        const startTime = process.hrtime.bigint();
        
        for (let i = 0; i < queryCount; i++) {
            const bid = orderbook.bestBid();
            const ask = orderbook.bestAsk();
        }
        
        const endTime = process.hrtime.bigint();
        const duration = Number(endTime - startTime) / 1000000000;
        
        const result = new BenchmarkResult('Price Queries', queryCount * 2, duration);
        this.results.push(result);
        return result;
    }

    /**
     * Benchmark mixed operations (realistic trading scenario)
     */
    benchmarkMixedOperations() {
        console.log('📊 Benchmarking mixed operations...');
        
        const orderbook = new JsOrderBook('BENCHMARK_MIXED');
        const operationCount = 5000;
        
        if (global.gc) global.gc();
        
        const startTime = process.hrtime.bigint();
        
        for (let i = 0; i < operationCount; i++) {
            // 70% order insertion, 30% price queries
            if (i % 10 < 7) {
                const price = 50000 + (i % 100);
                const side = i % 2 === 0 ? 'Buy' : 'Sell';
                orderbook.addLimitOrder(price, 100, side, 'GTC');
            } else {
                const bid = orderbook.bestBid();
                const ask = orderbook.bestAsk();
            }
        }
        
        const endTime = process.hrtime.bigint();
        const duration = Number(endTime - startTime) / 1000000000;
        
        const result = new BenchmarkResult('Mixed Operations', operationCount, duration);
        this.results.push(result);
        return result;
    }

    /**
     * Simulate concurrent access patterns
     */
    benchmarkConcurrentSimulation() {
        console.log('📊 Benchmarking concurrent simulation...');
        
        const orderbook = new JsOrderBook('BENCHMARK_CONCURRENT');
        const operationsPerWorker = 1000;
        const workerCount = 4;
        
        if (global.gc) global.gc();
        
        const startTime = process.hrtime.bigint();
        
        // Simulate concurrent access with interleaved operations
        for (let round = 0; round < operationsPerWorker; round++) {
            for (let worker = 0; worker < workerCount; worker++) {
                const price = 50000 + worker * 1000 + (round % 100);
                const side = round % 2 === 0 ? 'Buy' : 'Sell';
                orderbook.addLimitOrder(price, 100, side, 'GTC');
            }
        }
        
        const endTime = process.hrtime.bigint();
        const duration = Number(endTime - startTime) / 1000000000;
        const totalOperations = operationsPerWorker * workerCount;
        
        const result = new BenchmarkResult('Concurrent Simulation', totalOperations, duration);
        this.results.push(result);
        return result;
    }

    /**
     * Benchmark memory usage patterns
     */
    benchmarkMemoryUsage() {
        console.log('📊 Benchmarking memory usage...');
        
        const initialMemory = process.memoryUsage();
        
        const orderbook = new JsOrderBook('BENCHMARK_MEMORY');
        const orderCount = 50000;
        
        if (global.gc) global.gc();
        
        const startTime = process.hrtime.bigint();
        
        for (let i = 0; i < orderCount; i++) {
            const price = 50000 + (i % 1000);
            const side = i % 2 === 0 ? 'Buy' : 'Sell';
            orderbook.addLimitOrder(price, 100, side, 'GTC');
        }
        
        const endTime = process.hrtime.bigint();
        const duration = Number(endTime - startTime) / 1000000000;
        
        if (global.gc) global.gc();
        
        const finalMemory = process.memoryUsage();
        const memoryUsage = (finalMemory.heapUsed - initialMemory.heapUsed) / 1024 / 1024; // MB
        
        const result = new BenchmarkResult('Memory Usage', orderCount, duration, memoryUsage);
        this.results.push(result);
        return result;
    }

    /**
     * Benchmark latency distribution for individual operations
     */
    benchmarkLatencyDistribution() {
        console.log('📊 Benchmarking latency distribution...');
        
        const orderbook = new JsOrderBook('BENCHMARK_LATENCY');
        const operationCount = 10000;
        const latencies = [];
        
        // Warm up
        for (let i = 0; i < 100; i++) {
            orderbook.addLimitOrder(50000 + i, 100, 'Buy', 'GTC');
        }
        
        if (global.gc) global.gc();
        
        for (let i = 0; i < operationCount; i++) {
            const start = process.hrtime.bigint();
            const price = 50000 + (i % 1000);
            const side = i % 2 === 0 ? 'Buy' : 'Sell';
            orderbook.addLimitOrder(price, 100, side, 'GTC');
            const end = process.hrtime.bigint();
            latencies.push(Number(end - start) / 1000); // microseconds
        }
        
        // Calculate statistics
        latencies.sort((a, b) => a - b);
        const meanLatency = latencies.reduce((a, b) => a + b, 0) / latencies.length;
        const medianLatency = latencies[Math.floor(latencies.length / 2)];
        const p95Latency = latencies[Math.floor(0.95 * latencies.length)];
        const p99Latency = latencies[Math.floor(0.99 * latencies.length)];
        
        console.log(`    Mean latency: ${meanLatency.toFixed(2)} μs`);
        console.log(`    Median latency: ${medianLatency.toFixed(2)} μs`);
        console.log(`    95th percentile: ${p95Latency.toFixed(2)} μs`);
        console.log(`    99th percentile: ${p99Latency.toFixed(2)} μs`);
        
        const totalDuration = latencies.reduce((a, b) => a + b, 0) / 1000000; // convert to seconds
        const result = new BenchmarkResult('Latency Distribution', operationCount, totalDuration);
        this.results.push(result);
        return result;
    }

    /**
     * Run all benchmark tests
     */
    runAllBenchmarks() {
        console.log('🚀 Starting OrderBook Node.js Binding Benchmarks');
        console.log('='.repeat(60));
        
        const benchmarks = [
            this.benchmarkSingleOrderInsertion,
            this.benchmarkBatchOrderInsertion,
            this.benchmarkPriceQueries,
            this.benchmarkMixedOperations,
            this.benchmarkConcurrentSimulation,
            this.benchmarkMemoryUsage,
            this.benchmarkLatencyDistribution
        ];
        
        for (const benchmark of benchmarks) {
            try {
                const result = benchmark.call(this);
                console.log(`✅ ${result.name}: ${result.opsPerSecond.toLocaleString()} ops/sec`);
                if (result.memoryUsage !== 0) {
                    console.log(`    Memory usage: ${result.memoryUsage.toFixed(2)} MB`);
                }
            } catch (error) {
                console.log(`❌ ${benchmark.name} failed: ${error.message}`);
            }
        }
        
        this.printSummary();
        this.saveResults();
    }

    /**
     * Print benchmark summary
     */
    printSummary() {
        console.log('\n' + '='.repeat(60));
        console.log('📊 BENCHMARK SUMMARY');
        console.log('='.repeat(60));
        
        for (const result of this.results) {
            const name = result.name.padEnd(30, '.');
            const ops = result.opsPerSecond.toLocaleString().padStart(15);
            console.log(`${name} ${ops} ops/sec`);
        }
        
        if (this.results.length > 0) {
            const totalOps = this.results.reduce((sum, r) => sum + r.operations, 0);
            const totalTime = this.results.reduce((sum, r) => sum + r.duration, 0);
            const averageThroughput = totalOps / totalTime;
            
            console.log('-'.repeat(60));
            const avgName = 'Overall Average'.padEnd(30, '.');
            const avgOps = averageThroughput.toLocaleString().padStart(15);
            console.log(`${avgName} ${avgOps} ops/sec`);
            console.log(`${'Total Operations'.padEnd(30, '.')} ${totalOps.toLocaleString().padStart(15)} ops`);
            console.log(`${'Total Time'.padEnd(30, '.')} ${totalTime.toFixed(3).padStart(15)} sec`);
        }
    }

    /**
     * Save benchmark results to JSON file
     */
    saveResults() {
        const resultsFile = 'benchmark_results_nodejs.json';
        
        const resultsData = {
            timestamp: new Date().toISOString(),
            platform: 'Node.js',
            nodeVersion: process.version,
            results: this.results.map(r => r.toObject())
        };
        
        fs.writeFileSync(resultsFile, JSON.stringify(resultsData, null, 2));
        console.log(`\n📄 Results saved to ${resultsFile}`);
    }

    /**
     * Compare results with baseline performance
     */
    compareWithBaseline(baselineFile = 'benchmark_baseline.json') {
        if (!fs.existsSync(baselineFile)) {
            console.log(`⚠️  Baseline file ${baselineFile} not found`);
            return;
        }

        try {
            const baselineData = JSON.parse(fs.readFileSync(baselineFile, 'utf8'));
            
            console.log(`\n📈 PERFORMANCE COMPARISON vs ${baselineFile}`);
            console.log('='.repeat(60));
            
            const baselineResults = {};
            baselineData.results.forEach(r => {
                baselineResults[r.name] = r;
            });
            
            for (const result of this.results) {
                if (baselineResults[result.name]) {
                    const baselineOps = baselineResults[result.name].opsPerSecond;
                    const currentOps = result.opsPerSecond;
                    
                    if (baselineOps > 0) {
                        const improvement = ((currentOps - baselineOps) / baselineOps) * 100;
                        const status = improvement > 0 ? '📈' : improvement < -5 ? '📉' : '📊';
                        const name = result.name.padEnd(30, '.');
                        const perf = `${improvement.toFixed(1)}%`.padStart(8);
                        console.log(`${name} ${status} ${perf}`);
                    } else {
                        console.log(`${result.name.padEnd(30, '.')} ❓ No baseline data`);
                    }
                } else {
                    console.log(`${result.name.padEnd(30, '.')} 🆕 New benchmark`);
                }
            }
        } catch (error) {
            console.log(`❌ Error comparing with baseline: ${error.message}`);
        }
    }
}

/**
 * Main benchmark execution
 */
function main() {
    try {
        // Enable garbage collection if --expose-gc flag is used
        if (typeof global.gc === 'function') {
            console.log('🗑️  Garbage collection enabled for accurate memory measurements');
        } else {
            console.log('⚠️  Run with --expose-gc flag for accurate memory measurements');
        }
        
        const benchmark = new OrderBookBenchmark();
        benchmark.runAllBenchmarks();
        benchmark.compareWithBaseline();
        
        console.log('\n🎉 Node.js binding benchmarks completed successfully!');
        return 0;
        
    } catch (error) {
        console.log(`❌ Benchmark failed: ${error.message}`);
        return 1;
    }
}

// Run if this file is executed directly
if (require.main === module) {
    process.exit(main());
}

module.exports = { OrderBookBenchmark, BenchmarkResult };